src/lib.rs

```diff
@@ -527,14 +530,21 @@ impl OciDir {
 
     /// Find the manifest with the provided tag
     pub fn find_manifest_with_tag(&self, tag: &str) -> Result<Option<oci_image::ImageManifest>> {
-        let f = self.dir.open("index.json")?;
-        let idx: oci_image::ImageIndex = serde_json::from_reader(BufReader::new(f))?;
-        for img in idx.manifests() {
-            if Self::descriptor_is_tagged(img, tag) {
-                return self.read_json_blob(img).map(Some);
-            }
-        }
-        Ok(None)
+        let desc = self.find_manifest_descriptor_with_tag(tag)?;
+        desc.map(|img| self.read_json_blob(&img)).transpose()
+    }
+
+    /// Find the manifest descriptor with the provided tag
+    pub fn find_manifest_descriptor_with_tag(
+        &self,
+        tag: &str,
+    ) -> Result<Option<oci_image::Descriptor>> {
+        let idx = self.read_index()?.ok_or_else(|| Error::MissingImageIndex)?;
```

Comment:
Though shouldn't it be `read_index()` that's doing this error conversion?

Reply:
I didn't want to change the existing API but since we're going to bump the major version here's what I could do: change `read_index` to return `Result<ImageIndex>` instead of `Result<Option<ImageIndex>>`; the callers would have to handle the `MissingImageIndex` error instead of handling the possible `None` value returned by the existing API; does this sound reasonable?

Reply:
Hmmm...wait do we create an index by default? I forget...if we do, then I think we can unconditionally treat its absence as an error. Yeah looking at [the spec](https://github.com/opencontainers/image-spec/blob/main/image-layout.md) we have "It MUST exist" so this argues for dropping the `Option`.

Reply:
Interesting, I had the impression it was optional. I wonder why it says the following in the [Image index specification](https://github.com/opencontainers/image-spec/blob/main/image-index.md):
> The image index is a higher-level manifest which points to specific [image manifests](https://github.com/opencontainers/image-spec/blob/main/manifest.md), ideal for one or more platforms. While the use of an image index is OPTIONAL for image providers, image consumers SHOULD be prepared to process them.

Reply:
It is a bit confusing; AIUI the image index has a kind of dual role - served by a registry to do multiarch, but then it's also used for OCI image layouts to manage tags (and multiarch).